onnxruntime/core/platform/windows/env.cc

```diff
@@ -212,52 +221,121 @@ class WindowsEnv : public Env {
-      std::iota(ret.begin(), ret.end(), 0);
+    return std::max(1, phys_cores);
+  }
+
+  std::vector<ThreadOptions::ThreadAffinity> GetThreadAffinityMasks() const override {
+    // Default mask generation if everything else fails.
+    auto generate_vector_of_n = [](unsigned int num_logical_proc) {
+      // The return values will be converted into a 64-bit
+      // mask at the site of the SetThreadAffinityMask() call, so we need to make sure
+      // that num_logical_proc does not exceed 64, which is the max number of processors
+      // supported by SetThreadAffinityMask(). We assume two logical processors per
+      // physical core.
+      num_logical_proc = std::clamp(num_logical_proc, 1U, 64U);
+      std::vector<ThreadOptions::ThreadAffinity> ret;
+      if (num_logical_proc == 1U) {
+        ThreadOptions::ThreadAffinity th_aff{{0}};
+        ret.push_back(std::move(th_aff));
+        return ret;
+      }
+      ret.reserve(num_logical_proc / 2);
+      for (unsigned int c = 0; c < num_logical_proc; c += 2) {
+        ThreadOptions::ThreadAffinity th_aff{{static_cast<int>(c), static_cast<int>(c + 1)}};
+        ret.push_back(std::move(th_aff));
+      }
       return ret;
     };
-    // Indeed 64 should be enough. However, it's harmless to have a little more.
-    SYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer[256];
-    DWORD returnLength = sizeof(buffer);
-    if (GetLogicalProcessorInformation(buffer, &returnLength) == FALSE) {
+
+    std::vector<ThreadOptions::ThreadAffinity> ret;
+
+    auto logical_processor_info = FetchLogicalProcessorInfo();
+    if (!logical_processor_info.has_value()) {
       return generate_vector_of_n(std::thread::hardware_concurrency());
     }
-    std::vector<size_t> ret;
-    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
-    for (int i = 0; i != count; ++i) {
+
+    SYSTEM_LOGICAL_PROCESSOR_INFORMATION* buffer =
+        reinterpret_cast<SYSTEM_LOGICAL_PROCESSOR_INFORMATION*>(logical_processor_info->buffer_.get());
+
+    // Convert mask to a vector of ints
+    auto mask_to_vector = [](uint64_t mask) {
+      ThreadOptions::ThreadAffinity aff;
+      int bit = 0;
+      while (mask != 0) {
+        if ((mask & 0x1) != 0) {
+          aff.logical_proc_ids.push_back(bit);
+        }
+        mask >>= 0x1;
+        ++bit;
+      }
+      return aff;
+    };
+
+    const size_t count = logical_processor_info->count_;
+    ret.reserve(count);
+    for (size_t i = 0; i < count; ++i) {
       if (buffer[i].Relationship == RelationProcessorCore) {
-        ret.push_back(buffer[i].ProcessorMask);
+        // A single core can host multiple logical processors
+        // so the mask returned can have more than one bit set.
+        // In the past customers complained that threads jump from
+        // one logical process to another, we want to take just one
+        // bit per physical core.
+        ret.push_back(mask_to_vector(buffer[i].ProcessorMask));
```

Comment:
>ret.push_back(mask_to_vector(buffer[i].ProcessorMask)); [](http://example.com/codeflow?start=8&length=55)

So for windows, we are converting mask to vector here and vector to mask later ...

Reply:
thinking wildly - may be for windows we save mask in ThreadOptions::ThreadAffinity and use as it as mask as well.